### ISSUE TITLE
'Couldn't connect to server' at end of Step 4

### ISSUE THREAD

**Issue by sslccaj** [dylanjcastillo/fastapi-nginx-gunicorn#1]
Having trouble with:

curl --unix-socket unix:/home/fastapi-user/fastapi-nginx-gunicorn/run/gunicorn.sock http://localhost:8000

at the end of Step 4: Configure Gunicorn. Getting a response of:

curl: (7) Couldn't connect to server.

The command in the original terminal:

sudo supervisorctl status fastapi-app

shows the app as RUNNING. I'm not sure if it's a permissions thing, or if the gunicorn.sock is set up right?

**Comment by dylanjcastillo** [comment]
Hi @sslccaj,

A few things to investigate:

1. Did you create the `run` folder? 
2. Do you see a `gunicorn.sock` file in there? 
3. Do you see anything in the `gunicorn-error.log`?

Best,
Dylan

**Comment by sslccaj** [comment]
Thank you for the quick response Dylan!
Yes to points 1 and 2: there is a gunicorn.sock file in the run folder.
At the moment (4th complete restart with a new virtual environment each time), the answer to point 3 is no.
In previous renditions it has had errors, but it was while I was playing with "what if I change this" scenarios...

Best,
Sam

**Comment by dylanjcastillo** [comment]
Hmm, interesting. Can you share your `gunicorn_start` file and project structure (incl. subdirs if possible?

Also, the curl from the previous step worked right? (When running uvicorn main:app)

**Comment by sslccaj** [comment]
Hopefully this reply isn't public facing (I'm not very familiar with github). I hope this info helps.

Best,
Sam

Working:
[image.png]

Not working:
[image.png]

Project structure:
[image.png]
[image.png]
[image.png]
[image.png]
[image.png]
I put this file in because supervisor makes them with root permissions if the file doesn't already exist and I was wondering it that was causing some of my issues while I was trying out different "solutions".
[image.png]

[image.png]
[image.png]
[image.png]

**Comment by sslccaj** [comment]
Apparently these responses are public, but screenshots don't show (making my previous response rather useless). So I'll try again : )

Working:
uvicorn main:app
gives appropriate curl response on http://localhost:8000

Not working:
sudo supervisorctl status fastapi-app
shows RUNNING, but curl response is:
curl: (7) Couldn't connect to server

Project structure:
home
     user
          app			                rwxr-xr-x	        user
               _pycache_		        rwxr-xr-x	        user
               logs		                        rwxr-xr-x	        user
                    gunicorn-error.log	rw-r--r--	        user
               run		                        rwxr-xr-x	        user
                    gunicorn.sock	        rwxrwxrwx	user
               gunicorn_start		        rwxr--r--	        user
               main.py		                rw-r--r--	        user
               requirements.txt		rw-r--r--	        user
etc
     supervisor
          conf.d			                rwxr-xr-x	       root
               fastapi-app.conf		rw-r--r--	       root

Re: main.py
Exactly like github file

Re: requirements.txt (calls newer fastapi)
fastapi==0.91.0
gunicorn==20.1.0
uvicorn[standard]==0.20.0

Re: gunicorn-error.log
I put this file in because supervisor makes them with root permissions if the file doesn't already exist and I was wondering it that was causing some of my issues while I was trying out different "solutions".

Re: gunicorn_start
As follows (except for replacing user name with 'user'):

#!/bin/bash

NAME=pct-app
DIR=/home/user/app
USER=user
GROUP=user
WORKERS=3
WORKER_CLASS=uvicorn.workers.UvicornWorker
VENV=$DIR/.venv/bin/activate
BIND=unix:$DIR/run/gunicorn.sock
LOG_LEVEL=error

cd $DIR
source $VENV

exec gunicorn main:app \
  --name $NAME \
  --workers $WORKERS \
  --worker-class $WORKER_CLASS \
  --user=$USER \
  --group=$GROUP \
  --bind=$BIND \
  --log-level=$LOG_LEVEL \
  --log-file=-

**Comment by dylanjcastillo** [comment]
Hey @sslccaj,

Sorry, just figured out the issue. There was a small typo in the `curl` command.

This one should work:

```shell 
curl --unix-socket /home/ubuntu/fastapi-nginx-gunicorn/run/gunicorn.sock localhost
```

Changes:

1. `unix:` shouldn't have been there.
2. There was a `/` instead of an `-` in the project's directory name, so I fixed it.
3. I changed to `localhost` for brevity, but the previous one would also work.

Make sure to adjust the path to match your project's structure.

Thanks a lot for providing all this info.

Best,
Dylan
